### PR TITLE
DEV-2869 | Show alert when non-promise error when confirming Stripe payment

### PR DESCRIPTION
**Pull request by benjaminEwhite**
Please fill out each section even if it's just with "N/A"

#### What's this PR do?

- Catches non-promise errors (like the one that is captured in Sentry issue on the JIRA ticket for this PR) and displays an alert to user letting them know something has gone wrong.
- Adds tests for Stripe confirm payment errors
- Opportunistically removes unused, old Cypress commands

#### Why are we doing this? How does it help us?

Better UX in case of unexpected errors

#### How should this be manually tested? Please include detailed step-by-step instructions.

Not sure it can be. We've only seen one instance of the non-promise error that caused this, and not known how to reproduce.

#### Did this PR make changes to the user interface that may require changes to the user-facing docs?

No.

#### Have automated unit tests been added? If not, why?

Yes.

#### How should this change be communicated to end users?

N/A

#### Are there any smells or added technical debt to note?

No.

#### Has this been documented? If so, where?

N/A

#### What are the relevant tickets? Add a link to any relevant ones.

- [DEV-2869](https://news-revenue-hub.atlassian.net/browse/DEV-2869)
- [DEV-2921](https://news-revenue-hub.atlassian.net/browse/DEV-2921)

#### Do any changes need to be made before deployment to production (adding environment variables, for example)?

No

#### Are there next steps? If so, what? Have tickets been opened for them? List next-step tickets here:

[DEV-2921](https://news-revenue-hub.atlassian.net/browse/DEV-2921) -- there's a `console.error` call that should be updated after [DEV-2342](https://news-revenue-hub.atlassian.net/browse/DEV-2342) lands.

### CODE REVIEW

File: spa/cypress/e2e/01-donationPage.cy.js [hunk 1]
```diff
@@ -806,6 +807,66 @@ describe('User flow: unhappy paths', () => {
   });
 });
 
+describe('StripePaymentForm unhappy paths', () => {
+  beforeEach(() => {
+    // We intercept requests for Google Recaptcha because in test env, sometimes the live recaptcha returns an error
+    // (possibly because we load it too many times successively???), which was causing test failure else where.
+    cy.intercept({ method: 'GET', url: 'https://www.google.com/recaptcha/*' }, { statusCode: 200 });
```
Comment (nrh-cklimas) [inline spa/cypress/e2e/01-donationPage.cy.js:814]: Does it make sense to do this globally? I see these requests pop up all the time in Cypress.

File: spa/cypress/e2e/01-donationPage.cy.js [hunk 2]
```diff
@@ -806,6 +807,66 @@ describe('User flow: unhappy paths', () => {
...
+    cy.get('[data-testid*="frequency-one_time"]').click();
+    fillOutDonorInfoSection();
+    fillOutAddressSection();
+    fillOutReasonForGiving();
+    cy.get('form')
+      .findByRole('button', { name: /Continue to Payment/ })
+      .click();
+    cy.wait('@create-one-time-payment');
+  });
+  specify('when there is an unexpected but promise resolved error on Stripe payment element submission', () => {
```
Comment (nrh-cklimas) [inline spa/cypress/e2e/01-donationPage.cy.js:834]:
```suggestion
  specify('displays an error when there is an unexpected error, but the Stripe payment element submission payment promise resolved', () => {
```

File: spa/cypress/e2e/01-donationPage.cy.js [hunk 3]
```diff
@@ -806,6 +807,66 @@ describe('User flow: unhappy paths', () => {
...
+        amount: 123.01,
+        currencySymbol: livePageOne.currency.symbol,
+        frequency: CONTRIBUTION_INTERVALS.ONE_TIME
+      })
+    }).click();
+    cy.findByRole('alert').within(() => {
+      cy.contains(STRIPE_ERROR_MESSAGE).should('be.visible');
+    });
+  });
+  specify('There is an unexpected, non-promise error on Stripe payment element submission', () => {
```
Comment (nrh-cklimas) [inline spa/cypress/e2e/01-donationPage.cy.js:851]:
```suggestion
  specify('displays an error when an unexpected, non-promise error on Stripe payment element submission occurs', () => {
```

File: spa/cypress/e2e/01-donationPage.cy.js [hunk 4]
```diff
@@ -806,6 +807,66 @@ describe('User flow: unhappy paths', () => {
...
+      }
+    ).as('create-one-time-payment');
+    cy.visitDonationPage();
+    cy.get('[data-testid*="amount-120"]').click();
+    cy.get('[data-testid*="frequency-one_time"]').click();
+    fillOutDonorInfoSection();
+    fillOutAddressSection();
+    fillOutReasonForGiving();
+    cy.get('form')
+      .findByRole('button', { name: /Continue to Payment/ })
```
Comment (nrh-cklimas) [inline spa/cypress/e2e/01-donationPage.cy.js:830]:
```suggestion
      .findByRole('button', { name: 'Continue to Payment' })
```

If we need a loose match on this, I'd vote for `exact: false` in the function call.

File: spa/cypress/support/commands.js [hunk 5]
```diff
@@ -1,7 +1,6 @@
 import 'cypress-localstorage-commands';
 import '@testing-library/cypress/add-commands';
 
-import { TOKEN } from 'ajax/endpoints';
```
Comment (nrh-cklimas) [inline spa/cypress/support/commands.js:4]: :+1: for cleanup in this file

File: spa/src/components/paymentProviders/stripe/StripePaymentForm.js [hunk 6]
```diff
@@ -66,18 +68,29 @@ function StripePaymentForm() {
...
-      alert.error('An unexpected error occurred');
+    try {
+      const { error } = await stripe.confirmPayment({
+        elements,
+        confirmParams: { return_url, payment_method_data: { billing_details: stripeBillingDetails } }
+      });
+      // The Stripe docs note that this point will only be reached if there is an
+      // immediate error when confirming the payment. Otherwise, contributor gets redirected
+      // to `return_url` before the promise above ever resolves.
+      if (!['card_error', 'validation_error'].includes(error?.type)) {
```
Comment (nrh-cklimas) [inline spa/src/components/paymentProviders/stripe/StripePaymentForm.js:79]: Do these types of errors get handled elsewhere? I don't quite get this if statement.

**Review by nrh-cklimas — CHANGES_REQUESTED** [review]

File: spa/cypress/e2e/01-donationPage.cy.js (hunk 1 above)
Comment (benjaminEwhite) [reply spa/cypress/e2e/01-donationPage.cy.js]: It sounds like there's not any good options for global intercepts in Cypress, and if we end up with a test that is around recaptcha at some point, then we'd need to remember that recaptcha is being globally intercepted.

That said, it's mildly helpful to give this intercept a name, so I added a `cy.interceptGoogleRecaptcha` command, and replaced all of the recaptcha intercepts with this command.

4d032a2818e

Comment (nrh-cklimas) [reply spa/cypress/e2e/01-donationPage.cy.js]: Ah, right. I remember looking for this when I was trying to see if we could cause unintercepted requests to fail a test.

**Review by nrh-cklimas — APPROVED** [review]

**Review by x110dc — APPROVED** [review]

**Pull request merged**